add-chain/addresses.go

```diff
@@ -142,68 +142,94 @@ func readAddressesFromJSON(contractAddresses map[string]string, deploymentsDir s
 		ProxyAdmin,
 	}
 
-	contractsFromJSONFaultProofs := append(contractsFromJSON, []string{
+	customGasTokenContracts := []string{
+		AnchorStateRegistryProxy,
+		DelayedWETHProxy,
+		DisputeGameFactoryProxy,
+		MIPS,
+		PreimageOracle,
+		L2OutputOracleProxy,
+	}
+
+	faultProofContracts := []string{
 		AnchorStateRegistryProxy,
 		DelayedWETHProxy,
 		DisputeGameFactoryProxy,
 		FaultDisputeGame,
 		MIPS,
 		PermissionedDisputeGame,
 		PreimageOracle,
-	}...)
-	contractsFromJSONNonFaultProofs := append(contractsFromJSON, L2OutputOracleProxy)
-
-	contracts := contractsFromJSONFaultProofs
+	}
 
-	deployFilePath := filepath.Join(deploymentsDir, ".deploy")
-	_, err := os.Stat(deployFilePath)
+	contracts := universalContracts
 
+	// Check for the following
+	// 1. filepath == deploymentsDir
+	// 2. filepath == deploymentsDir/.deploy
+	// 3. filepath == deploymentsDir/<contract-name>.json
+	deployFilePath := filepath.Join(deploymentsDir)
+	fileInfo, err := os.Stat(deployFilePath)
 	if err != nil {
-		fmt.Printf("failed to find .deploy file. Will look for legacy .json files")
-		// Use legacy deployment artifact schema
+		return fmt.Errorf("invalid deployment filepath provided: %w", err)
+	}
 
-		_, err := os.ReadFile(filepath.Join(deploymentsDir, AnchorStateRegistryProxy+".json"))
-		if errors.Is(err, os.ErrNotExist) {
-			contracts = contractsFromJSONNonFaultProofs
-		}
-		for _, name := range contracts {
-			path := filepath.Join(deploymentsDir, name+".json")
-			file, err := os.ReadFile(path)
-			if err != nil {
-				return fmt.Errorf("failed to read file: %w", err)
+	if fileInfo.IsDir() {
+		deployFilePath = filepath.Join(deploymentsDir, ".deploy")
+		_, err = os.Stat(deployFilePath)
+		if err != nil {
+			// Use legacy deployment artifact schema
+			fmt.Printf("failed to find .deploy file. Will look for legacy .json files")
+
+			if _, err := os.ReadFile(filepath.Join(deploymentsDir, FaultDisputeGame+".json")); errors.Is(err, os.ErrNotExist) {
+				contracts = append(contracts, faultProofContracts...)
+			} else if _, err := os.ReadFile(filepath.Join(deploymentsDir, AnchorStateRegistryProxy+".json")); errors.Is(err, os.ErrNotExist) {
+				contracts = append(contracts, customGasTokenContracts...)
+			} else {
+				contracts = append(contracts, L2OutputOracleProxy)
 			}
```

Comment:
I'm a bit confused here. So if the `FaultDisputeGame+".json"` can _not_ be found, the FP contracts are appended to the `contracts` slice? And then again, if the `AnchorStateRegistryProxy+".json"` can _not_ be found, the CGT contracts are appended? Was the idea to negate `!` the `error.Is` checks? This would also better match the corresponding code below in [lines 217+](https://github.com/ethereum-optimism/superchain-registry/pull/425/files#diff-c8e7fef18ff265070f062af3aa6f25275760db61b16aac008aaca33c55e201acR217-R223).

Also, are FPs and CGTs mutually exclusive?

Reply:
Good catch. That logic is backwards. We want to use the FP contract set if `FaultDisputeGame.json` CAN be found.

AFAIK a chain using FP cannot use CGT. At least for right now. Can you confirm this is correct @sbvegan?

Reply:
Yeah my understanding is CGT chains do not support fault proofs. cc @tynes [to confirm ](https://github.com/ethereum-optimism/specs/discussions/140#discussioncomment-10191608)

Reply:
confirmed that this is the case at the moment, once https://github.com/ethereum-optimism/optimism/pull/10780 is merged then we will support cgt with fp

Reply:
Instead of hardcoding the list of expected addresses to find in the .deploy (or individual .json files), https://github.com/ethereum-optimism/superchain-registry/pull/425/commits/2df905bb646a753ae3ef2c35d0f75865ee341df4 improves the `readAddressesFromJSON` function to read all available addresses. This means we no longer have to maintain that list and the logic around it as features and expected contract sets evolve. We can rely on downstream add-chain and validation tests to check for the existence of required contracts. The same commit also adds a test case for the legacy deployment file type (one .json file for each contract instead of all in a single .deploy file). Previously that was untested code.

Reply:
Nice improvement!
> We can rely on downstream add-chain and validation tests to check for the existence of required contracts.

So this is guaranteed for _every_ contract?

Reply:
This is not guaranteed for every contract, but we do want to allow frontier chains to join the registry with non-standard contract sets. We rely on validation tests, especially the [standard tests](https://github.com/ethereum-optimism/superchain-registry/blob/main/validation/validation_test.go#L47-L63) to verify that contracts are correctly configured based on our expectations of the features on a standard chain.